### ISSUE TITLE
Conversion of submitter-supplied gene identifiers to Ensembl gene IDs: Other Affymetrix platforms

### ISSUE THREAD

**Issue by jaclyn-taroni** [AlexsLemonade/refinebio#329]
### Context

A subset of what was originally envisioned for #186. As of the writing of this ticket/#307, we add support for the detection of submitter-supplied gene identifiers and conversion to Ensembl gene IDs. Currently, this requires us to generate and store a mapping between the identifiers we will support conversion from and the Ensembl gene ids (see: https://github.com/AlexsLemonade/refinebio/pull/307#issuecomment-396364904). We have only generated this data for Brainarray-supported Affymetrix.

### Problem or idea

We'll need to generate this mapping for [non-Brainarray Affymetrix platforms we support](https://github.com/AlexsLemonade/refinebio/blob/e9900fa42459ab1b06e8980f5d8849705351f338/config/supported_microarray_platforms.csv).

### Solution or next step

I think the first step is determining the total number of samples that are on these platforms so we can figure out how to maximize our efforts (see https://github.com/AlexsLemonade/refinebio/issues/223#issuecomment-398118603), as I suspect some of these may require us to identify individual resources used by the organism in question's research community.

### New Issue Checklist

- [x] The title is short and descriptive.
- [x] You have explained the context that led you to write this issue.
- [x] You have reported a problem or idea.
- [x] You have proposed a solution or next step.

**Comment by jaclyn-taroni** [comment]
Depends on #223 platform info

**Comment by Miserlou** [comment]
The GEO-sourced `.db` files for the ones which match that pattern (which are actually most of them) are here: https://github.com/Miserlou/gene-converter/tree/master/geo

I think it only adds up to a few hundred samples total though. 

I think we should also exclude the ones that we can't process/noop otherwise, nor find Affymetrix-supplied conversion databases (`clarioms*`). This is only about a hundred samples.

**Comment by jaclyn-taroni** [comment]
Okay, just looking at GEO sample counts using the data over on https://github.com/AlexsLemonade/refinebio/issues/220, we get:


platform | sample_count_geo
-- | --
hthgu133pluspm | 8918
htmg430a | 3266
rhesus | 1877
bovgene10st | 340
chigene10st | 317
zebgene10st | 214
zebgene11st | 212
porgene11st | 206
bovgene11st | 183
porgene10st | 178
clariomsmouse | 177
cangene10st | 176
clariomshuman | 136
equgene10st | 107
clariomshumanht | 103
elegene10st | 78
ovigene11st | 60
ovigene10st | 37
chigene11st | 35
drogene11st | 34
elegene11st | 24
drogene10st | 21
cangene11st | 8
clariomsrat | 6
equgene11st | 6

We'll want to deal with at least `hthgu133pluspm` and `htmg430a`

**Comment by jaclyn-taroni** [comment]
Here's how we could go about it for `hthgu133pluspm`:
* Using the CEL file from [this sample](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM673476)
* Using the GPL file from [this platform accession](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL13158)

```R
# adapted from the brainarray stuff!

`%>%` <- dplyr::`%>%`

cel_path <- "GSM673476_SJHYPO008_D_HTU133PM.CEL"
platform <- "hthgu133pluspm"
org_code <- "hs"

# Does platform in CEL header match platform passed as an argument?
header_platform <- unlist(affyio::read.celfile.header(cel_path)[1])
header_platform <- tolower(gsub("[[:punct:]]", "", header_platform))
# if not, throw error
if (header_platform != platform) {
  stop("CEL file header does not match platform argument")
}

# load brainarray package
ba_package_name <- paste0(platform, org_code, "ensgprobe")
library(ba_package_name, character.only = TRUE)

# load bioconductor .db package
db_package_name <- paste0(db, ".db")
library(db_package_name, character.only = TRUE)

# generate a ExpressionFeatureSet from inputFile -- this step will automatically
# install platform design packages from bioconductor if not already installed
# adapted from SCAN.UPC one color
message("Reading..")
affy_feature_set <- oligo::read.celfiles(cel_path)
# this data.frame has the x, y coords and the manufacturer ID
probe_info_df <- oligo::getProbeInfo(affy_feature_set, 
                                     field = c("x", "y"), 
                                     probeType = "pm")

# data.frame contains x, y coord and the gene IDs from brainarray package
message("Mutating..")
brainarray_df <- as.data.frame(get(ba_package_name), stringsAsFactors = FALSE)
# we'll drop the trailing "_at" to get the Ensembl gene IDs
brainarray_df <- brainarray_df %>%
  dplyr::mutate(Probe.Set.Name = sub("_at", "", Probe.Set.Name))

# join brainarray and probe info by x,y coord & clean up a bit
message("Joining..")
ba_probe_ensg_df <- dplyr::inner_join(probe_info_df, brainarray_df,
                                      by = c("x", "y")) %>%
  dplyr::select(c("man_fsetid", "Probe.Set.Name")) %>%
  dplyr::mutate(PROBEID = man_fsetid, ENSEMBL = Probe.Set.Name) %>%
  dplyr::select(PROBEID:ENSEMBL) %>%
  dplyr::mutate(ENSEMBL = as.character(ENSEMBL))

# read in GPL file from geo and wrangle
gpl_df <- readr::read_tsv("GPL13158-5065.txt", comment = "#") %>%
  dplyr::select(c("ID", "Gene Symbol", "ENTREZ_GENE_ID", "RefSeq Transcript ID")) %>%
  dplyr::mutate(PROBEID = ID, 
                SYMBOL = `Gene Symbol`, 
                ENTREZID = ENTREZ_GENE_ID,
                REFSEQ = `RefSeq Transcript ID`) %>%
  dplyr::select(PROBEID:REFSEQ)

# some of the gene ids are separated by " /// " -- which columns have that?
sep_cols <- names(gpl_df)[apply(gpl_df, 2, function(x) any(grepl(" /// ", x)))]

# for each of the gene ids with multiple mappings, separate
for (col_name in sep_cols) {
  gpl_df <- gpl_df %>%
    tidyr::separate_rows(rlang::UQ(rlang::sym(col_name)), sep = " /// ")
}

# master
master_df <- dplyr::left_join(x = ba_probe_ensg_df, y = gpl_df, by = "PROBEID")
```

**Comment by jaclyn-taroni** [comment]
I believe we can do largely the same thing for `htmg430a` using [`GPL8759`](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL8759) with one slight change because the Entrez ID column name is different:

```R
# read in GPL file from geo and wrangle
gpl_df <- readr::read_tsv(<GPL8759_FILE>, comment = "#") %>%
  dplyr::select(c("ID", "Gene Symbol", "Entrez Gene", "RefSeq Transcript ID")) %>%
  dplyr::mutate(PROBEID = ID, 
                SYMBOL = `Gene Symbol`, 
                ENTREZID = `Entrez Gene`,
                REFSEQ = `RefSeq Transcript ID`) %>%
  dplyr::select(PROBEID:REFSEQ)
```

**Comment by Miserlou** [comment]
Handled here: https://github.com/AlexsLemonade/identifier-refinery/commit/5a07d20b3673fce8867c384e924c4a484c9a870d

**Comment by Miserlou** [comment]
Done
